src/libopensc/pkcs15-sc-hsm.c

```diff
@@ -862,6 +864,7 @@ static int sc_pkcs15emu_sc_hsm_init (sc_pkcs15_card_t * p15card)
 	appinfo->aid = sc_hsm_aid;
 
 	appinfo->ddo.aid = sc_hsm_aid;
+	sc_pkcs15_free_app(p15card);
```

Comment:
I assume some other initialization routine initialized this, but failed in the end (`sc_pkcs15_bind_internal()`?) This routine would be a better place for cleaning this up, wouldn't it?

Reply:
Actually, this one was more shot in the dark. Correct fix was 31e442f3 as you propose in pkcs15. I will drop ddea3c73 and 7d40797 as they do not seem to be much helpful.